crates/wysiwyg/src/composer_model/mentions.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright 2023 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{
+    dom::{DomLocation, Range},
+    ComposerModel, ComposerUpdate, DomNode, Location, SuggestionPattern,
+    UnicodeString,
+};
+
+impl<S> ComposerModel<S>
+where
+    S: UnicodeString,
+{
+    /// Remove the suggestion text and then insert a mention into the composer, using the following rules
+    /// - Do not insert a mention if the range includes link or code leaves
+    /// - If the composer contains a selection, remove the contents of the selection
+    /// prior to inserting a mention at the cursor.
+    /// - If the composer contains a cursor, insert a mention at the cursor
+    pub fn insert_mention_at_suggestion(
+        &mut self,
+        url: S,
+        text: S,
+        suggestion: SuggestionPattern,
+        attributes: Vec<(S, S)>,
+    ) -> ComposerUpdate<S> {
+        if self.should_not_insert_mention() {
+            return ComposerUpdate::keep();
+        }
+
+        self.replace_text_in(S::default(), suggestion.start, suggestion.end);
+        self.state.start = Location::from(suggestion.start);
+        self.state.end = self.state.start;
+
+        self.push_state_to_history();
+        self.do_insert_mention(url, text, attributes)
+    }
+
+    /// Inserts a mention into the composer. It uses the following rules:
+    /// - Do not insert a mention if the range includes link or code leaves
+    /// - If the composer contains a selection, remove the contents of the selection
+    /// prior to inserting a mention at the cursor.
+    /// - If the composer contains a cursor, insert a mention at the cursor
+    pub fn insert_mention(
+        &mut self,
+        url: S,
+        text: S,
+        attributes: Vec<(S, S)>,
+    ) -> ComposerUpdate<S> {
+        if self.should_not_insert_mention() {
+            return ComposerUpdate::keep();
+        }
+
+        if self.has_selection() {
+            self.replace_text(S::default());
+        }
+
+        self.push_state_to_history();
+        self.do_insert_mention(url, text, attributes)
```

Comment:
Same here

Reply:
To be honest sometimes I wonder if we should wrap this `ComposerModel` into another object that manages the undo stack and has all the public/FFI APIs, because it seems there are a lot very easy ways to push multiple states to history when we shouldn't.